New features: store form data, better validation

keflavich opened the pull request:
With this PR, all data submitted in the form will be included in the pull
request in a separate JSON file.  This will allow us to completely reconstruct
any data submission later.  That means we can allow other types of metadata
to be submitted and validated, even if we don't require it.

This change required a corresponding change in the unit validation scheme to
make it more flexible & powerful.

Reviews would be welcome!

keflavich committed:
store form data as json so that additional metadata can be handled in
the future

keflavich committed:
some minor bugfixes for form data storage

keflavich committed:
work toward equivalency validation (but this is a WIP because my testing
only works when I can access jquery, which means I need an internet
connection, which is not generally available on this plane...)

keflavich committed:
significantly improved validation: make sure the assigned unit is
equivalent to the column type
Also, search locally for jquery (which we'll store in the repo next
commit)

keflavich committed:
track jquery 1.10.2.min for offline testing

agianne commented:
Looks good to me.

e-koch commented:
:+1:

keflavich committed:
add a rathborne test (with a highly modified ecsv file) and fix bugs in
the upload_form revealed by running the test

keflavich committed:
update database & uploads (this can be done on master too)

keflavich committed:
fix tiny typo in error message

keflavich commented:
@e-koch @agianne thanks.  This is a pretty big change, so could you give it a test locally?  It would actually help to re-upload one of the data sets you've already uploaded so we can include the JSON data for those sets.

By test locally, I mean run the server locally and go all the way through the upload process.

e-koch commented on test_upload_form.py:
```diff
@@ -57,3 +57,90 @@ def test_upload_file(filename='benoitcommercon.csv',
 
     return S,r1,r2,r3
+
+def test_upload_rathborne(filename='rathborne2009_table2.ecsv',
```
Should we include this dataset (or a portion) for use in these unit tests? Same with 'benoitcommercon.csv'.

e-koch commented on upload_form.py:
```diff
@@ -107,10 +107,19 @@
 
 
+# Load the metadata unit mapping
+with open('alternate_allowed_metadata.json') as f:
```
@keflavich - Is this json file supposed to be included in the package or is it generated somewhere else?

keflavich committed:
add alternate metadata json file

keflavich committed:
add symlinks to the test upload files that are stored in uploads/

keflavich commented:
@e-koch thanks! I fixed both those issues.

agianne commented:
@keflavich If I try to reupload my table I get this error:

> Traceback (most recent call last):
>  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1836, in **call**
>    return self.wsgi_app(environ, start_response)
>  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1820, in wsgi_app
>    response = self.make_response(self.handle_exception(e))
>  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1403, in handle_exception
>    reraise(exc_type, exc_value, tb)
>  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1817, in wsgi_app
>    response = self.full_dispatch_request()
>  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1477, in full_dispatch_request
>    rv = self.handle_user_exception(e)
>  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1381, in handle_user_exception
>    reraise(exc_type, exc_value, tb)
>  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1475, in full_dispatch_request
>    rv = self.dispatch_request()
>  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1461, in dispatch_request
>    return self.view_functions[rule.endpoint](**req.view_args)
>  File "/home/andrea/Software/CAMELOT/frontend/upload_form.py", line 567, in set_columns
>    link_pull_database=link_pull_database,
>  File "/usr/local/lib/python2.7/dist-packages/flask/templating.py", line 128, in render_template
>    context, ctx.app)
>  File "/usr/local/lib/python2.7/dist-packages/flask/templating.py", line 110, in _render
>    rv = template.render(context)
>  File "/usr/local/lib/python2.7/dist-packages/jinja2/environment.py", line 969, in render
>    return self.environment.handle_exception(exc_info, True)
>  File "/usr/local/lib/python2.7/dist-packages/jinja2/environment.py", line 742, in handle_exception
>    reraise(exc_type, exc_value, tb)
>  File "/home/andrea/Software/CAMELOT/frontend/templates/show_plot.html", line 1, in top-level template code
>    {% extends 'base.html' %}
>  File "/home/andrea/Software/CAMELOT/frontend/templates/base.html", line 78, in top-level template code
>    {% block body %}
>  File "/home/andrea/Software/CAMELOT/frontend/templates/show_plot.html", line 76, in block "body"
>    {% if imagename[:-3] == 'png' %}
> TypeError: 'NoneType' object has no attribute '__getitem__'

but the PRs were created anyway!

keflavich commented:
@agianne OK, thanks - this indicates some sort of error with the plotter.  I don't understand it right away.  It looks like the variable "imagename" is never passed to `show_plot.html` from the traceback; could you investigate this?  Any clues would be helpful.  It may be a separate issue; do you get the same if you try uploading to the master camelot?

keflavich committed:
fix test data and tests: they need to be different from what's already
on the server.  Added a 'test mode' so that it is clear when these tests
are generating PRs as compared to real data sets.

agianne commented:
@keflavich Doing the same from the app results in 

> No difference was detected between the input branch of uploads/ ard the master after staging. >Possibly the submitted data file contains no data. 

I will investigate why these problems arise.

keflavich commented:
@agianne I can also reproduce the error, but I still don't understand why it's occurring.  I'm investigating.

keflavich committed:
additional testing code and related refactoring.  Provide a more
graceful failure mode if there are no data to plot.  Most important,
correctly set the column units!

keflavich commented:
The latest commit should allow you to run the tests, but please do so with caution - each time the test is run, it creates a new commit and e-mails anyone subscribed to the uploads and database repositories.  I've added a new `testmode='skip'` option to allow you to skip the pull request phase.

keflavich commented:
@agianne I believe the error you saw should be fixed now.

keflavich committed:
update benoitcommercon file so that it always is different from the one
on-server

e-koch commented on upload_form.py:
```diff
@@ -459,8 +550,11 @@ def set_columns(filename, fileformat=None):
+        if testmode:
+            raise ex
+        else:
+            return render_template('error.html', error=str(ex),
```
@keflavich - Maybe tack on a `, None` for when an error occurs. On line 495, a ValueError is raised (ie. Too many values to unpack) instead of linking to the error page.

e-koch replied:
The actual issue encountered was the duplicate checker, not a bug with this PR.

keflavich replied:
I'm not sure I understand what you're asking - is this a problem with the `render_template` command here?  If so, could you state the full command you think should be in place?  I'm not sure where `, None` belongs

e-koch replied:
Yeah, that wasn't very clear. The output is expected to be of the form to contain 2 items:

```
        link_pull_database, link_pull_uploads = \
                create_pull_request(username=username,
                                    merged_table=merged_table,
                                    merged_table_name=merged_table_name,
                                    table_widths=table_widths,
                                    unique_filename=unique_filename,
                                    testmode=testmode)
```

If the return for an error is:

```
            return render_template('error.html', error=str(ex),
                                   traceback=traceback.format_exc(ex))
```

A ValueError for "Too many values to unpack" is raised instead of creating the actual error page. The `, None` I meant would just get tacked on after the `render_template`.

keflavich replied:
Oh... right.  I hadn't appreciate that this change was returning `render_template` from a different function.  I'll come up with a solution.

agianne commented:
Ok, it now works locally!

keflavich committed:
rearrange the try/except clauses as per @e-koch's issue
warning: I don't know if try:/finally: works without an except.  I think
it does but I'm not 100% certain of the behavior

keflavich committed:
finally was dumb.  Different workaround now

e-koch commented:
The tests pass for me, but I get this error when trying to upload the ErikRosolowsky data:

```
Error: the remote URL git@github.com:camelot-project/uploads (which is really 'upl') does not match the expected one 'uploads'

Traceback (most recent call last): File "upload_form.py", line 502, in set_columns testmode=testmode) File "upload_form.py", line 560, in create_pull_request branch=branch_database, File "upload_form.py", line 626, in setup_submodule .format(check_upstream, database, name)) Exception: Error: the remote URL git@github.com:camelot-project/uploads (which is really 'upl') does not match the expected one 'uploads'
```

Is this just my local version?

agianne commented:
@e-koch What do you have in .gitmodules?

e-koch commented:
```
[submodule "database"]
        path = database
        url = git@github.com/camelot-project/database.git
[submodule "uploads"]
        path = uploads
        url = git@github.com/camelot-project/uploads.git
```

agianne commented:
@e-koch No, that's ok... We had a similar issue when the ".git" was missing from "url = git@github.com/camelot-project/uploads.git"

e-koch commented:
@agianne - I get the same error when trying to locally upload using the current master branch. So this is just something wonky in my local version.

@keflavich - Tests pass, so as far as I can tell it is good to go!

keflavich committed:
more verbose error message for @e-koch's issue

keflavich commented:
@e-koch I don't know exactly what's going on but I tried making the error message more verbose
